src/flint/fmpq.jl

```diff
@@ -288,97 +280,55 @@ conj(x::QQFieldElem) = x
 #
 ###############################################################################
 
-function +(a::QQFieldElem, b::QQFieldElem)
-  z = QQFieldElem()
-  add!(z, a, b)
-  return z
-end
++(a::QQFieldElem, b::QQFieldElem) = add!(QQFieldElem(), a, b)
 
-function -(a::QQFieldElem, b::QQFieldElem)
-  z = QQFieldElem()
-  sub!(z, a, b)
-  return z
-end
+-(a::QQFieldElem, b::QQFieldElem) = sub!(QQFieldElem(), a, b)
 
-function *(a::QQFieldElem, b::QQFieldElem)
-  z = QQFieldElem()
-  mul!(z, a, b)
-  return z
-end
+*(a::QQFieldElem, b::QQFieldElem) = mul!(QQFieldElem(), a, b)
 
 ###############################################################################
 #
 #   Ad hoc binary operators
 #
 ###############################################################################
 
-function +(a::QQFieldElem, b::Int)
-  z = QQFieldElem()
-  ccall((:fmpq_add_si, libflint), Nothing,
-        (Ref{QQFieldElem}, Ref{QQFieldElem}, Int), z, a, b)
-  return z
-end
-
-function +(a::QQFieldElem, b::ZZRingElem)
-  z = QQFieldElem()
-  ccall((:fmpq_add_fmpz, libflint), Nothing,
-        (Ref{QQFieldElem}, Ref{QQFieldElem}, Ref{ZZRingElem}), z, a, b)
-  return z
+for T in (ZZRingElem, Int, UInt)
+  @eval begin
+    +(a::QQFieldElem, b::$T) = add!(QQFieldElem(), a, b)
+    +(a::$T, b::QQFieldElem) = b + a
+  end
 end
 
-+(a::Int, b::QQFieldElem) = b + a
-
-+(a::ZZRingElem, b::QQFieldElem) = b + a
++(a::QQFieldElem, b::Integer) = a + ZZRingElem(b)
```

Comment:
Some idea in my head: dispatch this directly to add! and handle the conversion there (similar for the other arithmetics). Then the above loop can be removed and instead all methods be installed for IntegerUnion.

What do you think about this?

Reply:
I am OK with handling more in `add!`. However, we can't get rid of the loop because we'll get ambiguity errors (e.g. for `FracField + Int` vs `QQFieldElem + Integer`).

I suggest we merge this first and then we can think about handling additional cases. In fact I have another PR that helps with that in the pipeline.